// @antv/g2/lib/component/legend/tail.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * @fileOverview The class of tail legend
 * @author Ye Liu
 */

var Util = require('../../util');
var Category = require('./category');

var Tail = function (_Category) {
  _inherits(Tail, _Category);

  function Tail() {
    _classCallCheck(this, Tail);

    return _possibleConstructorReturn(this, _Category.apply(this, arguments));
  }

  Tail.prototype.getDefaultCfg = function getDefaultCfg() {
    var cfg = _Category.prototype.getDefaultCfg.call(this);
    return Util.mix({}, cfg, {
      /**
       * type标识
       * @type {String}
       */
      type: 'tail-legend',
      /**
       * 布局方式
       * horizontal 水平
       * vertical 垂直
       * @type {String}
       */
      layout: 'vertical'
    });
  };

  Tail.prototype._addItemMarker = function _addItemMarker(item, itemGroup) {
    var unCheckColor = this.get('unCheckColor');
    var markerAttrs = Util.mix({}, item.marker, {
      x: item.marker.radius,
      y: 0
    });

    if (!item.checked) {
      if (markerAttrs.fill) {
        markerAttrs.fill = unCheckColor;
      }
      if (markerAttrs.stroke) {
        markerAttrs.stroke = unCheckColor;
      }
    }

    var markerShape = itemGroup.addShape('marker', {
      type: 'marker',
      attrs: markerAttrs
    });
    markerShape.attr('cursor', 'pointer');
    markerShape.name = 'legend-marker';

    return markerShape;
  };

  Tail.prototype._addItemText = function _addItemText(item, itemGroup, startX) {
    var self = this;
    var unCheckColor = self.get('unCheckColor');
    var textStyle = self.get('textStyle');
    var textAttrs = Util.mix({}, textStyle, {
      x: startX,
      y: 0,
      text: self._formatItemValue(item.value)
    });
    if (!item.checked) {
      Util.mix(textAttrs, {
        fill: unCheckColor
      });
    }
    var textShape = itemGroup.addShape('text', {
      attrs: textAttrs
    });
    textShape.attr('cursor', 'pointer');
    textShape.name = 'legend-text';
    self.get('appendInfo') && textShape.setSilent('appendInfo', this.get('appendInfo'));
  };

  Tail.prototype._addItemWrapper = function _addItemWrapper(item, itemGroup, x, y) {
    var self = this;
    var bbox = itemGroup.getBBox();
    var itemWidth = self.get('itemWidth');
    var wrapperShape = itemGroup.addShape('rect', {
      attrs: {
        x: x,
        y: y - bbox.height / 2,
        fill: '#fff',
        fillOpacity: 0,
        width: itemWidth || bbox.width,
        height: bbox.height
      }
    });
    wrapperShape.attr('cursor', 'pointer');
    wrapperShape.setSilent('origin', item); // 保存图例项相关的数据，便于事件操作
    wrapperShape.name = 'legend-item';
    this.get('appendInfo') && wrapperShape.setSilent('appendInfo', this.get('appendInfo'));
  };

  Tail.prototype._addItem = function _addItem(item) {
    var self = this;
    var itemsGroup = self.get('itemsGroup');
    var x = 0;
    var y = 0;
    var itemGroup = itemsGroup.addGroup({
      x: x,
      y: y,
      value: item.value,
      scaleValue: item.scaleValue,
      checked: item.checked
    });
    itemGroup.translate(x, y);
    itemGroup.setSilent('viewId', itemsGroup.get('viewId'));
    var wordSpace = this.get('_wordSpaceing');
    var startX = 0;
    // 如果有marker添加marker
    if (item.marker) {
      var markerShape = self._addItemMarker(item, itemGroup);
      startX += markerShape.getBBox().width + wordSpace;
    }
    // text
    self._addItemText(item, itemGroup, startX);
    // 添加一个包围矩形，用于事件支持
    self._addItemWrapper(item, itemGroup, x, y);
    itemGroup.name = 'legendGroup';
    return itemGroup;
  };

  Tail.prototype._adjust = function _adjust() {
    var self = this;
    var geom = self.get('geom');
    if (geom) {
      var dataArray = self.get('geom').get('dataArray');
      var groups = this.get('itemsGroup').get('children');
      var index = 0;
      Util.each(groups, function (groupItem) {
        var dArray = dataArray[index];
        var lastY = dArray[dArray.length - 1].y;
        if (Util.isArray(lastY)) {
          lastY = lastY[1];
        }
        var groupHeight = groupItem.getBBox().height;
        var x = groupItem.get('x');
        groupItem.translate(x, lastY - groupHeight);
        index++;
      });
    }
  };

  Tail.prototype._renderUI = function _renderUI() {
    var self = this;
    if (!self.get('useHtml')) {
      // super._renderUI();
      self._renderItems();
      self.get('autoWrap') && self._adjustItems(); // 默认自动换行
      self._renderBack();
    } else {
      // 使用 html 渲染图例
      self._renderHTML();
    }
    var chart = self.get('chart');
    chart.once('afterpaint', function () {
      self._adjust();
    });
  };

  return Tail;
}(Category);

module.exports = Tail;
